File: run_validation.py
"""Run validation-test.html via Selenium and capture #REPORT output."""
import time
from selenium import webdriver
from selenium.webdriver.edge.service import Service
from selenium.webdriver.edge.options import Options
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from webdriver_manager.microsoft import EdgeChromiumDriverManager

def main():
    opts = Options()
    opts.add_argument("--headless=new")
    opts.add_argument("--disable-gpu")
    opts.add_argument("--no-sandbox")

    try:
        service = Service(EdgeChromiumDriverManager().install())
        driver = webdriver.Edge(service=service, options=opts)
    except Exception as e:
        print("Edge failed:", e, "- trying Chrome...")
        from selenium.webdriver.chrome.service import Service as CService
        from selenium.webdriver.chrome.options import Options as COpts
        from webdriver_manager.chrome import ChromeDriverManager
        co = COpts()
        co.add_argument("--headless=new")
        co.add_argument("--disable-gpu")
        co.add_argument("--no-sandbox")
        service = CService(ChromeDriverManager().install())
        driver = webdriver.Chrome(service=service, options=co)

    out_path = r"c:\Users\abahri\OneDrive - Cherryland Electric Cooperative\Desktop\Powerbuilding program\VALIDATION-OUTPUT.md"
    lines = []
    try:
        driver.get("http://localhost:3456/validation-test.html")
        def ready(d):
            el = d.find_element(By.ID, "REPORT")
            t = el.text.strip().lower()
            return "engine ready" in t or "run full" in t or "click" in t
        WebDriverWait(driver, 15).until(ready)
        time.sleep(1)
        driver.find_element(By.XPATH, "//button[contains(text(),'Run Full Validation')]").click()
        time.sleep(2)
        validation_text = driver.find_element(By.ID, "REPORT").text.strip()
        lines.append(validation_text)
        lines.append("\n\n---\n\n## STABILITY (NOISE) TESTS\n")
        driver.find_element(By.XPATH, "//button[contains(text(),'Run Stability')]").click()
        time.sleep(2)
        stability_text = driver.find_element(By.ID, "REPORT").text.strip()
        lines.append(stability_text)
        with open(out_path, "w", encoding="utf-8") as f:
            f.write("".join(lines))
        print("Saved to", out_path)
    except Exception as e:
        print("Error:", e)
        import traceback
        traceback.print_exc()
    finally:
        driver.quit()
    return 0

if __name__ == "__main__":
    exit(main())

File: run-validation.js
/**
 * Runs the validation suite in headless Chrome and saves output to VALIDATION-OUTPUT.md
 * Requires: npx puppeteer (run from Powerbuilding program folder)
 * Or start server first: npx serve -p 3456 -s
 */
const http = require('http');
const fs = require('fs');
const path = require('path');

const DIR = __dirname;
const PORT = 3456;

function serve(port) {
  return new Promise((resolve) => {
    const server = http.createServer((req, res) => {
      const file = req.url === '/' ? '/validation-test.html' : req.url;
      const filePath = path.join(DIR, file.replace(/^\//, ''));
      fs.readFile(filePath, (err, data) => {
        if (err) {
          res.writeHead(404);
          res.end('Not found');
          return;
        }
        const ext = path.extname(filePath);
        const ctypes = { '.html': 'text/html', '.js': 'application/javascript', '.css': 'text/css' };
        res.setHeader('Content-Type', ctypes[ext] || 'application/octet-stream');
        res.end(data);
      });
    });
    server.listen(port, () => resolve(server));
  });
}

async function runWithPuppeteer() {
  let puppeteer;
  try {
    puppeteer = require('puppeteer');
  } catch {
    console.log('Puppeteer not installed. Run: npm install puppeteer');
    process.exit(1);
  }

  let server;
  try {
    server = await serve(PORT);
  } catch (e) {
    console.log('Could not start server:', e.message);
    process.exit(1);
  }

  const browser = await puppeteer.launch({ headless: true });
  const page = await browser.newPage();

  try {
    await page.goto(`http://localhost:${PORT}/validation-test.html`, { waitUntil: 'networkidle0', timeout: 15000 });
    await page.waitForFunction(() => {
      const eng = document.getElementById('ENG')?.contentWindow;
      return eng && eng.ENGINE_READY;
    }, { timeout: 10000 });

    await page.click('button[onclick="runValidation()"]');
    await page.waitForTimeout(1500);

    const validationReport = await page.evaluate(() => document.getElementById('REPORT')?.textContent || '');

    await page.click('button[onclick="runStabilityTests()"]');
    await page.waitForTimeout(2000);

    const stabilityReport = await page.evaluate(() => document.getElementById('REPORT')?.textContent || '');

    const fullOutput = `# Engine Validation Report — Full Output

Generated: ${new Date().toISOString()}

---

## VALIDATION SUITE
${validationReport}

---

## STABILITY (NOISE) TESTS
${stabilityReport}

---

END
`;

    const outPath = path.join(DIR, 'VALIDATION-OUTPUT.md');
    fs.writeFileSync(outPath, fullOutput, 'utf8');
    console.log('Output saved to VALIDATION-OUTPUT.md');
  } catch (e) {
    console.error('Error:', e.message);
  } finally {
    await browser.close();
    server.close();
  }
}

runWithPuppeteer();
